invalid cgen of str methods for structs with more than 8 members

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, Ubuntu 20.04.3 LTS
Processor: 2 cpus, 64bit, little endian, Intel(R) Core(TM) i3-3225 CPU @ 3.30GHz
CC version: cc (Ubuntu 10.3.0-1ubuntu1~20.04) 10.3.0

getwd: /v/cleanv
vmodules: /home/delian/.vmodules
vroot: /v/cleanv
vexe: /v/cleanv/v
vexe mtime: 2022-07-23 12:20:36
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.0 6ebe668.d8b0df1

Git version: git version 2.25.1
Git vroot status: weekly.2022.29-52-gd11fcd37-dirty
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 827f7452

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg cgen_str_bug_triggered_by_structs_with_more_than_8_members.v`
```v
struct Abc {
	a string
	b string
	c string
	d string
	//
	e string
	f string
	g string
	h string
	//
	x int // number of fields must be > 8
}

// the custom str() method is needed; the auto generated one works
fn (a Abc) str() string {
	return ''
}

fn (a Abc) some_method() {
	print(a) // needed
}

```

**What did you expect to see?**

compiled program


**What did you see instead?**
```
==================
/tmp/v_1000/cgen_str_bug_triggered_by_structs_with_more_than_8_members.2511807054363903402.tmp.c:12104: error: cannot convert 'struct main__Abc' to 'struct main__Abc *'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/gen/c/str.v
+++ b/vlib/v/gen/c/str.v
@@ -106,7 +106,7 @@ fn (mut g Gen) gen_expr_to_string(expr ast.Expr, etype ast.Type) {
 		g.write('${str_fn_name}(')
 		if str_method_expects_ptr && !is_ptr {
 			g.write('&')
-		} else if (!str_method_expects_ptr && is_ptr && !is_shared) || is_var_mut {
+		} else if !str_method_expects_ptr && !is_shared && (is_ptr || is_var_mut) {
 			g.write('*')
 		}
 		if expr is ast.ArrayInit {

--- a/vlib/v/tests/custom_str_on_struct_with_too_many_fields_test.v
+++ b/vlib/v/tests/custom_str_on_struct_with_too_many_fields_test.v
@@ -1,0 +1,29 @@
+struct Abc {
+	a string
+	b string
+	c string
+	d string
+	//
+	e string
+	f string
+	g string
+	h string
+	//
+	x int // number of fields must be > 8
+}
+
+fn (a Abc) str() string {
+	return 'abc'
+}
+
+fn (a Abc) some_method() string {
+	println(a)
+	return '$a'
+}
+
+fn test_custom_str_on_struct_with_too_many_fields() {
+	abc := Abc{}
+	ret := abc.some_method()
+	println(ret)
+	assert ret == 'abc'
+}

